### PR TITLE
MARKET-1383 Markdown

### PR DESCRIPTION
**Pull request by StasivPavlo**
dark theme
light theme

### CODE REVIEW

File: components/bl-markdown-component/src/index.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+export default function MyCustomComponent({ component }) {
+  const { classList, style, display, url, markdownText, height, width } = component;
+
+  const [markdown, setMarkdown] = useState('');
+  const [isLoad, setIsLoad] = useState(false);
+  const [errorMessage, setErrorMessage] = useState('');
+
+  useEffect(() => {
+    const md = new MarkdownIt();
```
Comment (v-excelsior) [inline components/bl-markdown-component/src/index.js:14]: I'm not sure about creating new instance of library every changing of this properties.
Could it be created only once at first render?

File: components/bl-markdown-component/src/index.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+      fetch(url)
+        .then(response => response.text())
+        .then((data) => {
+          setIsLoad(false);
+          setMarkdown(md.render(data));
+        })
+        .catch((error) => {
+          setIsLoad(false);
+          setErrorMessage(error.message);
+        });
```
Comment (v-excelsior) [inline components/bl-markdown-component/src/index.js:26]: `setIsLoad(false);` must be at finally

File: components/bl-markdown-component/src/index.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  const { classList, style, display, url, markdownText, height, width } = component;
+
+  const [markdown, setMarkdown] = useState('');
+  const [isLoad, setIsLoad] = useState(false);
+  const [errorMessage, setErrorMessage] = useState('');
+
+  useEffect(() => {
+    const md = new MarkdownIt();
+    if (!markdownText && url) {
+      setIsLoad(true);
```
Comment (v-excelsior) [inline components/bl-markdown-component/src/index.js:16]: empty line

File: components/bl-markdown-component/src/index.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
+import MarkdownIt from './markdown-it.min.js';
+import { useState, useEffect } from 'react';
+
+const { cn } = BackendlessUI.CSSUtils;
+
+export default function MyCustomComponent({ component }) {
+  const { classList, style, display, url, markdownText, height, width } = component;
+
+  const [markdown, setMarkdown] = useState('');
+  const [isLoad, setIsLoad] = useState(false);
```
Comment (v-excelsior) [inline components/bl-markdown-component/src/index.js:10]: isLoading or loading

File: components/bl-markdown-component/src/index.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    <div
+      className={ cn('bl-customComponent-markdown markdown-body', classList) }
+      style={ { ...style, height: height || '100%', width: width || '100%' } }>
+      { isLoad && <Loader/> }
+
+      { errorMessage && !isLoad ? (
+        <span className="error-message"> { errorMessage } </span>
+      ) : (
+        <div className="markdown-body" dangerouslySetInnerHTML={ { __html: markdown } }/>
+      ) }
```
Comment (v-excelsior) [inline components/bl-markdown-component/src/index.js:48]: I'm not pretty love notations like this, they look a little dirty.

What about something like that?

```
const renderMdContent = () => {
   if(isLoading) {
     return <Loader>
   }
   
   if(errorMessage) {
     return <Error>
   }

  return <div className="markdown-body" .../>
}
```

And here 

```
return (
  <div>
    { renderMdContent() }
  </div>
)
```

It's just an example, you can refactor it using this principle

File: components/bl-markdown-component/component.json [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+      "dataBinding": true,
+      "handlerDescription": "This is a handler for the logic to determine height for component"
+    },
+    {
+      "type": "text",
+      "name": "width",
+      "label": "Width",
+      "showInSettings": true,
+      "hasLogicHandler": true,
+      "handlerId": "widthLogic",
```
Comment (v-excelsior) [inline components/bl-markdown-component/component.json:49]: handler label?

File: components/bl-markdown-component/component.json [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+      "handlerDescription": "This is a handler for the logic to determine text of markdown",
+      "dataBinding": true
+    },
+    {
+      "type": "text",
+      "name": "height",
+      "label": "Height",
+      "showInSettings": true,
+      "hasLogicHandler": true,
+      "handlerId": "heightLogic",
```
Comment (v-excelsior) [inline components/bl-markdown-component/component.json:39]: handler label?

File: components/bl-markdown-component/component.json [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+      "label": "Width",
+      "showInSettings": true,
+      "hasLogicHandler": true,
+      "handlerId": "widthLogic",
+      "dataBinding": true,
+      "handlerDescription": "This is a handler for the logic to determine width for component"
+    }
+  ],
+  "eventHandlers": [],
+  "actions": [],
```
Comment (v-excelsior) [inline components/bl-markdown-component/component.json:55]: add actions:

* set md content
* set md url

File: components/bl-markdown-component/src/index.js [hunk 9]
```diff
@@ -0,0 +1,62 @@
+import MarkdownIt from './markdown-it.min.js';
```
Comment (v-excelsior) [inline components/bl-markdown-component/src/index.js:1]: I'm not sure, but our eslint config has import order and relative imports should be under the libraries

**Review by v-excelsior — CHANGES_REQUESTED** [review]

File: components/bl-markdown-component/src/index.js (hunk 9 above)
Comment (StasivPavlo) [reply components/bl-markdown-component/src/index.js:3]: yes, I fixed it

File: components/bl-markdown-component/src/index.js [hunk 10]
```diff
@@ -0,0 +1,85 @@
+import { useEffect, useMemo, useState } from 'react';
+
+import MarkdownIt from './markdown-it.min.js';
+
+const { cn } = BackendlessUI.CSSUtils;
+
+export default function MyCustomComponent({ component }) {
```
Comment (v-excelsior) [inline components/bl-markdown-component/src/index.js:7]: Rename it

File: components/bl-markdown-component/src/index.js [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+  const { classList, style, display, url, markdownText, height, width } = component;
+
+  const [markdown, setMarkdown] = useState('');
+  const [isLoading, setIsLoading] = useState(false);
+  const [errorMessage, setErrorMessage] = useState('');
+
+  const mdUrl = useMemo(() => url, [url]);
```
Comment (v-excelsior) [inline components/bl-markdown-component/src/index.js:14]: Do we really need it?

File: components/bl-markdown-component/src/index.js [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+
+  component.setContent = (text) => {
+    loadMd(mdUrl, text);
+  };
+
+  useEffect(() => {
+    loadMd(mdUrl, markdownText);
+  }, [markdownText, mdUrl]);
+
+  const loadMd = (url, text) => {
```
Comment (v-excelsior) [inline components/bl-markdown-component/src/index.js:30]: Could we split it into two functions setContent/fetchContent or something like that?

File: components/bl-markdown-component/src/index.js [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+
+      fetch(url)
+        .then(response => response.text())
+        .then((data) => {
+          setMarkdown(md.render(data));
+          setIsLoading(false);
+        })
+        .catch((error) => {
+          setErrorMessage(error.message);
+          setIsLoading(false);
```
Comment (v-excelsior) [inline components/bl-markdown-component/src/index.js:42]: use `finaly` for `setIsLoading(false);`

File: components/bl-markdown-component/src/index.js [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+
+  if (!display) {
+    return null;
+  }
+
+  return (
+    <div
+      className={ cn('bl-customComponent-markdown markdown-body', classList) }
+      style={ { ...style, height: height || '100%', width: width || '100%' } }>
+      <MdComponent isLoading={ isLoading } errorMessage={ errorMessage } markdown={ markdown }/>
```
Comment (v-excelsior) [inline components/bl-markdown-component/src/index.js:59]: MdContent will be better i think

**Review by v-excelsior — CHANGES_REQUESTED** [review]

**Pull request merged**